app/validators/routing/new_when_rule_validator.py

```diff
@@ -0,0 +1,196 @@
+OPERATOR_LESS_THAN_OR_EQUAL = "<="
+OPERATOR_IN = "in"
+OPERATOR_ALL_IN = "all-in"
+OPERATOR_ANY_IN = "any-in"
+OPERATOR_DATE = "date"
+
+BOOLEAN_OPERATORS = [OPERATOR_NOT, OPERATOR_AND, OPERATOR_OR]
+
+COMPARISON_OPERATORS = [
+    OPERATOR_EQUAL,
+    OPERATOR_NOT_EQUAL,
+    OPERATOR_GREATER_THAN,
+    OPERATOR_LESS_THAN,
+    OPERATOR_GREATER_THAN_OR_EQUAL,
+    OPERATOR_LESS_THAN_OR_EQUAL,
+]
+
+ARRAY_OPERATORS = [OPERATOR_IN, OPERATOR_ALL_IN, OPERATOR_ANY_IN]
+
+VALUE_OPERATORS = [OPERATOR_DATE]
+
+ALL_OPERATORS = (
+    BOOLEAN_OPERATORS + COMPARISON_OPERATORS + ARRAY_OPERATORS + VALUE_OPERATORS
+)
+
+
+class NewWhenRuleValidator(Validator):
+    OPERATOR_ARGUMENT_TYPE_MISMATCH = "Argument types don't match"
+    INVALID_ARGUMENT_TYPE_FOR_OPERATOR = "Invalid argument type for operator"
+    VALUE_DOESNT_EXIST_IN_ANSWER_OPTIONS = "Value doesn't exist in answer options"
+    DATE_OPERATOR_REFERENCES_NON_DATE_ANSWER = (
+        "Date operator references non Date answer"
+    )
+
+    def __init__(self, when_clause, origin_id, questionnaire_schema):
+        super().__init__(when_clause)
+        self.when_clause = when_clause
+        self.questionnaire_schema = questionnaire_schema
+        self.context["origin_id"] = origin_id
+
+    def validate(self):
+        """
+        Validates all operators and arguments in when clause
+        """
+        self.validate_rule(self.when_clause)
+
+        return self.errors
+
+    def validate_rule(self, rule):
+        operator_name = next(iter(rule))
+        argument_types = []
+        for argument in rule[operator_name]:
+            if isinstance(argument, dict) and any(
+                operator in argument for operator in ALL_OPERATORS
+            ):
+                argument_type = self.validate_rule(argument)
+            elif isinstance(argument, dict) and "source" in argument:
+                argument_type = resolve_value_source_json_type(
+                    argument, self.questionnaire_schema.answers_with_context
+                )
+            else:
+                argument_type = python_type_to_json_type(type(argument).__name__)
+            argument_types.append(argument_type)
+
+        if operator_name == OPERATOR_DATE:
+            self._validate_date_operator(rule)
+
+        if operator_name in COMPARISON_OPERATORS + ARRAY_OPERATORS:
+            self._validate_comparison_operator_argument_types(
+                rule, operator_name, argument_types
+            )
+
+        if (
+            operator_name in COMPARISON_OPERATORS + [OPERATOR_ALL_IN, OPERATOR_ANY_IN]
+            and TYPE_NULL not in argument_types
+        ):
+            self._validate_argument_types_match(rule, argument_types)
+
+        if operator_name in ARRAY_OPERATORS + [OPERATOR_EQUAL, OPERATOR_NOT_EQUAL]:
+            self._validate_options(rule, operator_name)
+
+        if operator_name == OPERATOR_DATE:
+            return TYPE_DATE
+        return TYPE_BOOLEAN
```

Comment:
This is fun, like a mathematic guess who, but although a bit hard to digest does the job. Can't say I have gone through all combinations mind, which is a slight concern, might loop back later.

Reply:
This isn't meant to exhaustively validate all the operations. There could be an operator that has no need for any code validation.